File: o2a/quota/registry.py
"""额度适配器注册表 + auto 域名嗅探（§8.3）。

选择逻辑：accounts[].quota_source
- auto：按 openai_url 域名嗅探（openrouter.ai → openrouter；嗅探不到 → local）
- 显式名：openrouter / local / local-rolling-5h / manual / none
- 未注册的显式名（anthropic / codex / zen 等）→ 预留名，回退 local

失败隔离（§8.4-3）：fetch 抛错/超时 → 上层 get_snapshot 降级 local 并标 stale。
"""

import asyncio
import threading
from urllib.parse import urlparse

from .adapters.local import LocalQuotaAdapter
from .adapters.local_rolling_5h import LocalRolling5hAdapter
from .adapters.manual import ManualQuotaAdapter
from .adapters.openrouter import OpenRouterAdapter
from .base import QuotaAdapter, QuotaContext, QuotaError, make_snapshot

_ADAPTERS = {}


def register(adapter: QuotaAdapter):
    """注册一行：新增一个供应商适配 = 新增一个文件 + 这里一行（§8.4-1）。"""
    _ADAPTERS[adapter.name] = adapter


register(LocalQuotaAdapter())
register(LocalRolling5hAdapter())
register(ManualQuotaAdapter())
register(OpenRouterAdapter())

# 域名嗅探表：子串 → 适配器名（auto 用）
_SNIFF = [
    ("openrouter.ai", "openrouter"),
]

# 预留显式名（尚未实现 → 回退 local）
_RESERVED = {"anthropic", "codex", "openai_codex", "zen", "opencode_zen", "generic"}


def resolve_adapter_name(account) -> str:
    source = (getattr(account, "quota_source", "") or "auto").strip()
    if source in _RESERVED:
        return "local"
    if source != "auto" and source in _ADAPTERS:
        return source
    url = (getattr(account, "openai_url", "") or "")
    host = urlparse(url).netloc.lower()
    for frag, name in _SNIFF:
        if frag in host:
            return name
    # auto 且手填了额度上限（quota.limit）→ manual 套餐（冷启动兜底）
    quota = getattr(account, "quota", None)
    if source == "auto" and isinstance(quota, dict) and quota.get("limit"):
        return "manual"
    return "local"


def registered_adapters():
    return sorted(_ADAPTERS.keys())


_snapshot_cache = {}
_cache_lock = threading.Lock()


def _fetch_timeout(adapter, ctx: QuotaContext):
    result = adapter.fetch(ctx)
    return result


async def _finalize(snapshot, account, ctx: QuotaContext, name: str, ttl_cache=None, degrade=True):
    """降级 + 缓存写回（sync/async 共用）。"""
    if snapshot is None and degrade and name != "local":
        try:
            local = _ADAPTERS.get("local")
            if local is not None:
                snapshot = _fetch_timeout(local, ctx)
                if asyncio.iscoroutine(snapshot):
                    snapshot = await asyncio.wait_for(snapshot, timeout=UPSTREAM_TIMEOUT_S)
                if isinstance(snapshot, dict):
                    snapshot["stale"] = True
        except Exception:
            snapshot = None
    if snapshot is None and degrade:
        # local 也失败（如统计目录缺失）→ 最小快照，标 stale
        snapshot = make_snapshot("local", [], stale=True, now_fn=ctx.now_fn)
    if snapshot is not None and ttl_cache:
        ttl_cache.set(getattr(account, "id", "") or "", snapshot)
    return snapshot


async def get_snapshot_async(account, ctx: QuotaContext, ttl_cache=None, degrade=True):
    """取额度快照：优先缓存 → 注册表适配器 → 失败降级 local（标 stale）。

    适配器异常/超时绝不外泄（§8.4-3 失败隔离）。"""
    if account is None:
        return None
    key = getattr(account, "id", "") or ""
    if ttl_cache:
        cached = ttl_cache.get(key)
        if cached is not None:
            return cached
    name = resolve_adapter_name(account)
    snapshot = None
    adapter = _ADAPTERS.get(name)
    if adapter is not None:
        try:
            result = _fetch_timeout(adapter, ctx)
            if asyncio.iscoroutine(result):
                result = await asyncio.wait_for(result, timeout=UPSTREAM_TIMEOUT_S)
            snapshot = result
        except Exception:
            snapshot = None
    return await _finalize(snapshot, account, ctx, name, ttl_cache=ttl_cache, degrade=degrade)


def get_snapshot(account, ctx: QuotaContext, ttl_cache=None):
    """同步入口：local 系同步适配器零开销直调；async 适配器需事件循环。"""
    import asyncio

    if account is None:
        return None
    key = getattr(account, "id", "") or ""
    if ttl_cache:
        cached = ttl_cache.get(key)
        if cached is not None:
            return cached
    name = resolve_adapter_name(account)
    adapter = _ADAPTERS.get(name)
    if adapter is None or not asyncio.iscoroutinefunction(adapter.fetch):
        snapshot = None
        try:
            if adapter is not None:
                snapshot = _fetch_timeout(adapter, ctx)
                assert not asyncio.iscoroutine(snapshot)
        except Exception:
            snapshot = None
        return asyncio.run(_finalize(snapshot, account, ctx, name, ttl_cache=ttl_cache))
    try:
        loop = asyncio.get_running_loop()
    except RuntimeError:
        return asyncio.run(get_snapshot_async(account, ctx, ttl_cache=ttl_cache))
    else:
        return loop.run_until_complete(get_snapshot_async(account, ctx, ttl_cache=ttl_cache))

File: crates/o2a-quota/src/registry.rs
//! 额度适配器注册表 + auto 域名嗅探（对齐 Python `o2a/quota/registry.py`）。

use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;

use o2a_config::Account;
use serde_json::Value;
use url::Url;

use crate::adapters::{self, codex::OpenAICodexAdapter, declarative::DeclarativeQuotaAdapter, local::LocalQuotaAdapter, local_rolling_5h::LocalRolling5hAdapter, manual::ManualQuotaAdapter, opencode_go::OpenCodeGoAdapter, openrouter::OpenRouterAdapter, zai::ZaiAdapter};
use crate::base::{make_snapshot, QuotaContext, TTLCache, UPSTREAM_TIMEOUT_S};

/// 适配器协议：fetch(ctx) -> snapshot | None。
///
/// fetch 内允许抛 QuotaError / 任意错误 —— 注册表统一捕获降级。
/// Python 的同步 fetch 与 async fetch 双轨在此统一为 async（engine 为 tokio 异步）。
#[async_trait::async_trait]
pub trait QuotaAdapter: Send + Sync {
    fn name(&self) -> &'static str;
    fn source(&self) -> &'static str {
        "local_stats"
    }
    async fn fetch(&self, ctx: &QuotaContext) -> Result<Option<Value>, crate::base::QuotaError>;
}

/// 注册表：适配器按名字索引（新增一个供应商 = 新文件 + register 一行）。
pub struct Registry {
    adapters: HashMap<&'static str, Arc<dyn QuotaAdapter>>,
}

impl Default for Registry {
    fn default() -> Self {
        Registry::standard()
    }
}

impl Registry {
    /// 全部 8 个内置适配器（对齐 Python 模块级 register 调用）。
    pub fn standard() -> Self {
        let mut r = Registry {
            adapters: HashMap::new(),
        };
        r.register(Arc::new(LocalQuotaAdapter));
        r.register(Arc::new(LocalRolling5hAdapter));
        r.register(Arc::new(ManualQuotaAdapter));
        r.register(Arc::new(OpenRouterAdapter));
        r.register(Arc::new(DeclarativeQuotaAdapter));
        r.register(Arc::new(OpenCodeGoAdapter));
        r.register(Arc::new(ZaiAdapter));
        r.register(Arc::new(OpenAICodexAdapter));
        r
    }

    pub fn register(&mut self, adapter: Arc<dyn QuotaAdapter>) {
        self.adapters.insert(adapter.name(), adapter);
    }

    pub fn get(&self, name: &str) -> Option<Arc<dyn QuotaAdapter>> {
        self.adapters.get(name).cloned()
    }

    pub fn contains(&self, name: &str) -> bool {
        self.adapters.contains_key(name)
    }
}

/// 已注册适配器名（排序），对齐 `registered_adapters`。
pub fn registered_adapters() -> Vec<String> {
    Registry::standard()
        .adapters
        .keys()
        .map(|s| s.to_string())
        .collect()
}

// 域名嗅探表：子串 → 适配器名（auto 用）
const SNIFF: [(&str, &str); 6] = [
    ("openrouter.ai", "openrouter"),
    ("opencode.ai", "opencode-go"),
    ("chatgpt.com", "codex"),
    ("openai.com", "codex"),
    ("bigmodel.cn", "zai"),
    ("z.ai", "zai"),
];

// 显式名 → 适配器名别名（支持套餐名等用户友好写法）
const ALIASES: [(&str, &str); 9] = [
    ("glm-coding-plan", "zai"),
    ("glm", "zai"),
    ("codex_zen", "zai"),
    ("opencode_go", "opencode-go"),
    ("openai-codex", "codex"),
    ("openai_codex", "codex"),
    ("gpt", "codex"),
    ("chatgpt", "codex"),
    ("openai", "codex"),
];

// 预留显式名（尚未实现 → 回退 local）
const RESERVED: [&str; 3] = ["anthropic", "zen", "generic"];

/// 选择适配器名（对齐 `resolve_adapter_name`，含显式未注册名继续嗅探的语义）。
pub fn resolve_adapter_name(registry: &Registry, account: &Account) -> String {
    let mut source = account.quota_source.trim().to_string();
    if source.is_empty() {
        source = "auto".into();
    }
    if RESERVED.contains(&source.as_str()) {
        return "local".into();
    }
    let source = ALIASES
        .iter()
        .find(|(a, _)| *a == source)
        .map(|(_, b)| b.to_string())
        .unwrap_or(source);
    if source != "auto" && registry.contains(&source) {
        return source;
    }
    let url = &account.openai_url;
    let host = Url::parse(url)
        .map(|u| {
            let mut netloc = u.host_str().unwrap_or("").to_lowercase();
            if let Some(port) = u.port() {
                netloc.push_str(&format!(":{}", port));
            }
            netloc
        })
        .unwrap_or_default();
    for (frag, name) in SNIFF {
        if host.contains(frag) {
            return name.into();
        }
    }
    // auto 且手填了额度上限（quota.limit）→ manual 套餐（冷启动兜底）
    if source == "auto" {
        if let Some(quota) = &account.quota {
            if quota.get("limit").map(crate::base::truthy).unwrap_or(false) {
                return "manual".into();
            }
        }
    }
    "local".into()
}

/// 取额度快照：优先缓存 → 注册表适配器（1.5s 超时）→ 失败降级 local（标 stale）→
/// local 也失败 → 最小空窗口快照（标 stale）。适配器异常/超时绝不外泄。
///
/// degrade=false（对齐 Python `degrade` 参数）：不降级、不缓存，直接返回 None。
pub async fn get_snapshot_async(
    registry: &Registry,
    account: &Account,
    ctx: &QuotaContext,
    ttl_cache: Option<&TTLCache>,
    degrade: bool,
) -> Option<Value> {
    let key = account.id.clone();
    if let Some(cache) = ttl_cache {
        if let Some(cached) = cache.get(&key) {
            return Some(cached);
        }
    }
    let name = resolve_adapter_name(registry, account);
    let mut snapshot = None;
    if let Some(adapter) = registry.get(&name) {
        let fut = adapter.fetch(ctx);
        if let Ok(Ok(Some(s))) =
            tokio::time::timeout(Duration::from_secs_f64(UPSTREAM_TIMEOUT_S), fut).await
        {
            snapshot = Some(s);
        }
    }
    finalize(registry, snapshot, account, ctx, &name, ttl_cache, degrade).await
}

/// 降级 + 缓存写回（对齐 `_finalize`）。
async fn finalize(
    registry: &Registry,
    snapshot: Option<Value>,
    account: &Account,
    ctx: &QuotaContext,
    name: &str,
    ttl_cache: Option<&TTLCache>,
    degrade: bool,
) -> Option<Value> {
    let mut snapshot = snapshot;
    if snapshot.is_none() && degrade && name != "local" {
        if let Some(local) = registry.get("local") {
            let fut = local.fetch(ctx);
            if let Ok(Ok(Some(mut s))) =
                tokio::time::timeout(Duration::from_secs_f64(UPSTREAM_TIMEOUT_S), fut).await
            {
                if let Value::Object(o) = &mut s {
                    o.insert("stale".into(), Value::Bool(true));
                }
                snapshot = Some(s);
            }
        }
    }
    if snapshot.is_none() && degrade {
        // local 也失败（如统计目录缺失）→ 最小快照，标 stale
        snapshot = Some(make_snapshot(
            "local",
            Vec::new(),
            "local_stats",
            None,
            true,
            &ctx.now(),
        ));
    }
    if let (Some(s), Some(cache)) = (&snapshot, ttl_cache) {
        cache.set(&account.id.clone(), s.clone());
    }
    snapshot
}

// adapters 模块对外再导出，保持与 Python `o2a.quota.adapters` 相同的访问路径
pub use adapters::{
    codex, declarative, local, local_rolling_5h, manual, opencode_go, openrouter, zai,
};
